# <b>Streaming data and buffers</b>

Python io module allows us to manage the file-related input and output operations. The advantage of using the IO module is that the classes and functions available allows us to extend the functionality to enable writing to the Unicode data.


In [25]:
# Creating a byte types and example

mybinarytext=b'Hello World!'

type(mybinarytext)


bytes

In [26]:
#creating a empty 32bytes..


print(bytes(32))

#creating a 4bytes

x=b'\x00\x01\x02\x03'
print(x)

#Using an ASCII representaion

print('\xFF')

# coverting a value to an interger
print(ord(b'\xFF'))
print(type(ord(b'\xFF')))

# print the bytes indexes
y=b'\xf0\xff'
print(y[1],y[0],sep="\t")

#print out the ACSII bytes representation
print("my data bytes is:{:b}".format(y[0]), ",my data bytes is:{:b}".format(y[1]))
print("my data bytes is:{:d}".format(y[0]), ", my data bytes is:{:x}".format(y[1]))
print("my data bytes is:{:04x}".format(y[0]))
print("my data bytes is:{:08x}".format(y[1]))

#looping through the a binary number


b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
b'\x00\x01\x02\x03'
ÿ
255
<class 'int'>
255	240
my data bytes is:11110000 ,my data bytes is:11111111
my data bytes is:240 , my data bytes is:ff
my data bytes is:00f0
my data bytes is:000000ff


## <b>io.BytesIO()  Class</b>
A data stream as the name suggests is a stream of data. We can visualize this as the stream of water flowing from one end to another. 

Each droplet of water is the packet of data flowing in the stream and we can resize this data packet as per our needs.
If we have a massive file to deal with, we can process the file’s data one packet at a time instead of loading the entire file in the memory. Also, we can write to a file or data store one packet at a time.

Buffer, on the other hand, is a temporary region of volatile memory (RAM) where data can be stored before consuming it. We can read from or write to a data buffer using streams.


In [1]:
import io

In [2]:
bio=io.BytesIO()

type(bio)

_io.BytesIO

In [3]:
bio.write(b'Hello\n')

6

Writing to a buffer as buffer only under bytes, 
write string, number to buffer by encoding the data to ascii codes or Utf-8

In [4]:
bio.write("hello".encode("ascii"))

5

In [5]:
# The cursor is moved to the begining of data for reading
bio.seek(0)

0

This first 5 bytes of the data and when this called again it reads the next five bytes of the data

In [6]:
bio.read(5)

b'Hello'

In [7]:
bio.read(5)

b'\nhell'

And the out type of a buffer are immutable

In [17]:
type(_)

bytes

But if we want the data writen out of buffer accessible, then we use getbuffer()


In [20]:
bio.getbuffer()
type(bio.getbuffer())

memoryview

In [21]:
buff=bio.getbuffer()
buff[0]=255

In [22]:
bio.seek(0)

0

In [23]:
#Read out buffer mutable read returns a hexidecimal
bio.read(1)

b'\xff'

In [34]:
print(f'{255:08b},{255:08x},"\t"')
print(f'{15:08b},{15:08x},"\t"')
print(f'{1:08b},{1:08x},"\t"')

11111111,000000ff,"	"
00001111,0000000f,"	"
00000001,00000001,"	"


### <b> Bitwise Operators</b>

"""
EXAMPLE of binary calculation and Hexidecimal calculation
"""

- and         denoted by &(ampersand)
- OR.         denoted by |
- XOR.        Denoted by ^(Caret)
- Compliment  denoted by ~

### <b>AND => &</b>

<table>
    
 <tr>
        <th><b>X</b></th>
        <th><b>Y</b></th>
        <th><b>Result</b></th>
  </tr>
  <tr>
    <td>1</td>
    <td>0</td>
    <td>0</td>
   </tr>
<tr>
    <td>1</td>
    <td>0</td>
    <td>0</td>
   </tr
    
  <tr>
    <td>1</td>
    <td>1</td>
    <td>1</td>
 </tr>
</table>

### <b>OR => |</b>

<table>
    
 <tr>
        <th><b>X</b></th>
        <th><b>Y</b></th>
        <th><b>Result</b></th>
  </tr>
  <tr>
    <td>1</td>
    <td>0</td>
    <td>1</td>
   </tr>
<tr>
    <td>0</td>
    <td>1</td>
    <td>0</td>
   </tr>
  <tr>
    <td>1</td>
    <td>1</td>
    <td>1</td>
 </tr>
</table>


### <b>XOR => ^</b>

<table>
    
 <tr>
        <th><b>X</b></th>
        <th><b>Y</b></th>
        <th><b>Result</b></th>
  </tr>
  <tr>
    <td>1</td>
    <td>0</td>
    <td>1</td>
   </tr>
<tr>
    <td>0</td>
    <td>1</td>
    <td>1</td>
   </tr>
  <tr>
    <td>1</td>
    <td>1</td>
    <td>0</td>
 </tr>
</table>


### <b>Complimet => ^</b>

<table>
    
 <tr>
        <th><b>X</b></th>
        <th><b>Result</b></th>
  </tr>
  <tr>
    <td>1</td>
    <td>0</td>
   </tr>
<tr>
    <td>0</td>
    <td>1</td>
 
   </tr>
</table>



Performing Bitwise Calculation

In [3]:
def printbinary(x):
    x=print(f'{x:08b}')
    return x

####  <b>Note that only 255 makes a full binary bytes</b>

In [45]:
print(printbinary(15),printbinary(255),"\t")

00001111
11111111
None None 	


In [47]:
printbinary(255 & 15)

00001111


In [48]:
printbinary(255|15)

11111111


In [49]:
printbinary(255^15)

11110000


In [83]:
print(~255)
"""
This is called ones complimenent
- tidel performs a flip operations transforming postive to negative and vice-versa multiplying them by -1
- What happens here is 256*-1-1
Perform
"""

-256


'\nThis is called ones complimenent\n- tidel performs a flip operations transforming postive to negative and vice-versa multiplying them by -1\n- What happens here is 256*-1-1\nPerform\n'

### </b>Application of binary operations</b>
- A practical example of when you may actually use this is to check if a specific bit is turned

For example, you could pack a <b> boolean variables</b> into a single byte.

- Let's say you're making a video game and you need to store data efficiently.

- You could use one bit to represent whether it's daytime or nighttime.

- Another bit can be whether it's raining or not.

- Another bit can store whether a character is male or female.

- Another for hot or cold rain or snow and so on.

- If you want to check, if one specific bit is on, you can use the end operation to check that.

Assuming we want check weather the third is turned set as can see the third it is set

In [53]:
printbinary(4)

00000100


Perform AND bitwise operations with function above

In [55]:
printbinary(4&12)

00000100


In [57]:
printbinary(12)

00001100


In [59]:
printbinary(4&99)

00000000


In [62]:
printbinary(4&25)

00000000


In [63]:
printbinary(25)

00011001


<b> We could write a function that check if a bitwise operation returns 4 </b>

In [65]:
def is_thirsbit_set(x):
     return(x & 4)==4

In [66]:
is_thirsbit_set(12)

True

In [67]:
is_thirsbit_set(99)

False

<b>Performing OR bitwise operations</b>

In [71]:
printbinary(12)

00001100


In [72]:
printbinary(1)

00000001


In [73]:
printbinary(12|1)

00001101


In [74]:
printbinary(12|255)

11111111


<b>Perform XOR bitwise operations</b>

In [77]:
printbinary(12^255)

11110011


In [78]:
printbinary(1^12)

00001101


#### <b>Perform Bit shifting</b>

In [84]:
printbinary(15)

00001111


<b> Left Shifting</b> This potentially result in loss of data

In [85]:
printbinary(15<<1)

00011110


In [86]:
printbinary(15<<2)

00111100


In [87]:
printbinary(15<<3)

01111000


<b>Right Shifting </b> This acn potentiall result in loss of data

In [90]:
printbinary(255>>1)

01111111


In [91]:
printbinary(255>>2)

00111111


In [92]:
printbinary(255>>3)

00011111


### <b>CONVSERSION BETWEEN DATA TYPES</b>
- you have an integer or maybe you have a hexadecimal string that you need to convert into actual bytes.
- hexidecimal-FFA03A
- Binary string (1111010)
- Integer (25)
- Little endian architecture of a computer this starts from the bigest number to smallest 
<table>
    
 <tr>
        <th><b>33</b></th>
        <th><b>22</b></th>
        <th><b>11</b></th>
     <th><b>00</b></th>
  </tr></table>
 
 
- Big endian architecture of a computer starts from the bigest number to smallest 
<table>
    
 <tr>
        <th><b>00</b></th>
        <th><b>11</b></th>
        <th><b>22</b></th>
     <th><b>33</b></th>
  </tr></table>
  
And you need to shift between these datatypes to fit a Scenario

<table>
<tr>
        <th><b>datatype</b></th>
        <th><b>Description</b></th>
  </tr> 
<tr><td>byte	1 bytes</td> <td>	Stores whole numbers from -128 to 127</td> </tr>
<tr><td>short	2 bytes</td><td>	Stores whole numbers from -32,768 to 32,767 <td></tr>
<tr><td>int	4 bytes      </td><td>  Stores whole numbers from -2,147,483,648 to 2,147,483,647<td></tr>
<tr><td>long	8 bytes</td><td>	Stores whole numbers from -9,223,372,036,854,775,808 to 9,223,372,036,854,775,807<td></tr>
<tr><td>float	4 bytes</td><td>	Stores fractional numbers. Sufficient for storing 6 to 7 decimal digits<td></tr>
<tr><td>double	8 bytes</td><td>	Stores fractional numbers. Sufficient for storing 15 decimal digits<td></tr>
<tr><td>boolean	1 bit</td><td>	Stores true or false values<td></tr>
<tr><td>char	2 bytes</td><td>	Stores a single character/letter or ASCII values<td></tr>
</table>

In [5]:
#int to bytes
i=16
i.to_bytes(1,byteorder="big",signed=True)

b'\x10'

In [8]:
f=9000000000
f.to_bytes(8,byteorder="big",signed=True)

b'\x00\x00\x00\x02\x18q\x1a\x00'

<b>Python converts integer to ASCII characters </b>

In [9]:
bytes([65,64,70])

b'A@F'

In a 64bit system each interger takes 64bits which is equal 8bytes of memory


mylist=[255,254,253,252] # this list take quatilebytes of memory, but if stored in binary will consume on 4bits of memory

In [10]:
mylist=[255,254,253,252]
bytes(mylist)

b'\xff\xfe\xfd\xfc'

<b>Converting Binary String to Bytes</b>
String "11110000-> int:240-> Bytes:xF0

<table><tr><td>Base</td>
       <td>Description</td>
    <tr><td>Base 2:</td>  <td>Binary</td></tr>
<tr><td>Base 8:</td>  <td>Octal</td></tr>
<tr><td>Base 10: <td>Decimal</td></tr>
<tr><td>Base 16: <td>Hecadecimal </td></tr>

</table>

In [25]:
s="11110000"
# Converting binary string to int requires specifying the base
s=int(s,2)

#Coverting to bytes
s.to_bytes(1,byteorder="big",signed=False)

b'\xf0'

Alternatively

In [20]:
s="11110000"
bytes([int(s,2)])

b'\xf0'

<b>Converting Bytes to integer</b> Bytes:xF0->int:240


In [29]:
s=b'\xf0\xfff\xf0'

In [30]:
int.from_bytes(s,byteorder='big',signed=True)

-251697424

<b>Hex or Hexdecimal is way represent number in base 16 from 0-15 and F are used to repr num from 10-15 </b>

In [2]:
#suppose we have this complex and want to convert to bytes
s="DEADBEEF"
import binascii

binascii.unhexlify(s)

b'\xde\xad\xbe\xef'

converting  bytes to hexidecimal

In [3]:
binascii.hexlify(bytes([255,255,255,255]))

b'ffffffff'

In [9]:
x=binascii.hexlify(b'\x0f\xff')
x

b'0fff'

In [12]:
#decoding this 
x.decode("ascii")
x.decode("utf8")

'0fff'

 <b>Working with Files </b>
 Note when working with file 
 - to overwrite and write a new file use "wb"
 - to append "a+"
 - read a file "r"
 - write to a file w+ this will overwrite all file if exist
 - read and write binary r+b

In [23]:
import io
with open("test.dat","wb")  as myfile:
    myfile.write(b'\xff\xfe\xfd\xfc')


In [16]:
with open("test.dat","rb") as readfile:
          #for line in readfile.read(): This will return the decode binary in int on a \n
           #print(readfile.read())  # this will read and return the entire binary data in the file
            print(readfile.read(1)) # this will return specific index
         

b'\xff'


### <b>Getting Meta data of a file</b>

In [5]:
#checking if the file exists
import os

os.path.exists("test.dat")

True

In [6]:
# checking for the size of the file and possible exceptions
try: 
    print(os.path.getsize("test.dat"))
except FileNotFoundError:
    print("I can't find the file")

4


In [7]:
try: 
    print(os.path.getsize("tes.dat"))
except FileNotFoundError as fileerror:
    print(fileerror)

[Errno 2] No such file or directory: 'tes.dat'


In [8]:
# checking for file owner
os.stat('test.dat')

os.stat_result(st_mode=33188, st_ino=12886548560, st_dev=16777220, st_nlink=1, st_uid=502, st_gid=20, st_size=4, st_atime=1610650731, st_mtime=1610650730, st_ctime=1610650730)

- st_mode=file permissions
- st_uid= user identifer file owner
- st_gid= group identifer file owner
- st_atime= Time of the most recent access in seconds
- st_mt=time of most recent modification
- st_ctime=in windows the time of creation expr inseconds, in mac/unix time of most recent meta data change 
- st_dev= identifier of the device on whic this file resides
- st_size= file size in bytes

#### Iterating over file with a WHILE  and stopping when there is nothing to return

In [20]:
f=open("test.dat","rb")
bytes_read=f.read(1)
while bytes_read:
    print(bytes_read)
    bytes_read=f.read(1)
    

b'\xff'
b'\xfe'
b'\xfd'
b'\xfc'


#### Scenario of writing a data into file in bytes  and jumping to differrent section of the file

In [23]:
with open("test1.dat","wb") as file2:
    for i in range(255):
        file2.write(bytes([i]))
        


In [27]:
with open("test1.dat","rb") as file3:
     print(file3.read())
     
     print(os.stat("test1.dat"))
        
    

b'\x00\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x0b\x0c\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\x7f\x80\x81\x82\x83\x84\x85\x86\x87\x88\x89\x8a\x8b\x8c\x8d\x8e\x8f\x90\x91\x92\x93\x94\x95\x96\x97\x98\x99\x9a\x9b\x9c\x9d\x9e\x9f\xa0\xa1\xa2\xa3\xa4\xa5\xa6\xa7\xa8\xa9\xaa\xab\xac\xad\xae\xaf\xb0\xb1\xb2\xb3\xb4\xb5\xb6\xb7\xb8\xb9\xba\xbb\xbc\xbd\xbe\xbf\xc0\xc1\xc2\xc3\xc4\xc5\xc6\xc7\xc8\xc9\xca\xcb\xcc\xcd\xce\xcf\xd0\xd1\xd2\xd3\xd4\xd5\xd6\xd7\xd8\xd9\xda\xdb\xdc\xdd\xde\xdf\xe0\xe1\xe2\xe3\xe4\xe5\xe6\xe7\xe8\xe9\xea\xeb\xec\xed\xee\xef\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7\xf8\xf9\xfa\xfb\xfc\xfd\xfe'
os.stat_result(st_mode=33188, st_ino=12886555662, st_dev=16777220, st_nlink=1, st_uid=502, st_gid=20, st_size=255, st_atime=1610654449, st_mtime=1610654449, st_ctime=1610654449)


#### Seek function allows to jump around different part of the file
-  However if we want to jump to the end of the file we have to import from os SEEk_END,SEEk_CUR , 
-  first we have to specific our starting

In [31]:
with open("test1.dat","rb") as file3:
     print(file3.seek(0),file3.seek(1))
        

0 1


In [36]:
from os import SEEK_END,SEEK_CUR

with open("test1.dat","rb") as file3:
    print(file3.seek(0,SEEK_END))
 # we are doing here is moving cur to end of the data and starting from there 0-where to seek and SEEK_END- last part of the data
 # This can also be achieve with
    print(file3.seek(0,2))
 # if we try to read the file again this will return empty becos the cursor is at the nd of file
    print(file3.read())
#  However we cound read backwards
    print(file3.seek(-4,2))
# if we try to read the file again we will get avalue
    print(file3.read())

255
255
b''
251
b'\xfb\xfc\xfd\xfe'
